experimental/kernels/kernels_c.cpp

```diff
@@ -0,0 +1,525 @@
+#include "gpu.h"
+#include <array>
+#include <cstdio>
+#include <future>
+
+#include "kernels.h"
+#include "kernels_c.h"
+
+using namespace gpu; // createContext, createTensor, createKernel,
+                     // createShader, dispatchKernel, wait, toCPU
+                     // Tensor, Kernel, Context, Shape, kf32
+
+#define VOCAB_SIZE 50257
+
+// See https://github.com/google/dawn/blob/a8fbe981a86cb59536e2de423d2013a82d9b54a0/src/dawn/native/Limits.cpp
+#define LIMITS_BUFFER_SIZE_1GB { \
+    .limits = { \
+      .maxTextureDimension1D=8192, \
+      .maxTextureDimension2D=8192, \
+      .maxTextureDimension3D=2048, \
+      .maxTextureArrayLayers=256, \
+      .maxBindGroups=4, \
+      .maxBindGroupsPlusVertexBuffers=24, \
+      .maxBindingsPerBindGroup=1000, \
+      .maxDynamicUniformBuffersPerPipelineLayout=8, \
+      .maxDynamicStorageBuffersPerPipelineLayout=4, \
+      .maxSampledTexturesPerShaderStage=16, \
+      .maxSamplersPerShaderStage=16, \
+      .maxStorageBuffersPerShaderStage=8, \
+      .maxStorageTexturesPerShaderStage=4, \
+      .maxUniformBuffersPerShaderStage=12, \
+      .maxUniformBufferBindingSize=65536, \
+      .maxStorageBufferBindingSize=1073741824, \
+      .minUniformBufferOffsetAlignment=256, \
+      .minStorageBufferOffsetAlignment=256, \
+      .maxVertexBuffers=8, \
+      .maxBufferSize=0x80000000, \
+      .maxVertexAttributes=16, \
+      .maxVertexBufferArrayStride=2048, \
+      .maxInterStageShaderComponents=64, \
+      .maxInterStageShaderVariables=16, \
+      .maxColorAttachments=8, \
+      .maxColorAttachmentBytesPerSample=32, \
+      .maxComputeWorkgroupStorageSize=16384, \
+      .maxComputeInvocationsPerWorkgroup=256, \
+      .maxComputeWorkgroupSizeX=256, \
+      .maxComputeWorkgroupSizeY=256, \
+      .maxComputeWorkgroupSizeZ=64, \
+      .maxComputeWorkgroupsPerDimension=65535 \
+    }, \
+    .nextInChain = nullptr \
+  }
+
+void ENCODER_FORWARD_GPU(float* out,
+                         int* inp, float* wte, float* wpe,
+                         int B, int T, int C){
+  unsigned long b = static_cast<unsigned long>(B);
+  unsigned long t = static_cast<unsigned long>(T);
+  unsigned long c = static_cast<unsigned long>(C);
+  unsigned long v = VOCAB_SIZE;
+  struct EncoderParams {
+    uint32_t B;
+    uint32_t T;
+    uint32_t C;
+  };
+  setLogLevel(kError);
+  WGPURequiredLimits requiredLimits = LIMITS_BUFFER_SIZE_1GB;
+  Context ctx = createContext({},{},{
+      .requiredLimits = &requiredLimits
+    });
+  Tensor input = createTensor(ctx, Shape{b * t}, ki32, inp);
```

Comment:
Wouldn't u32 be a better fit for token IDs than i32 or is there something i might be missing?

Reply:
The type of token id in llm.c is int. It uses the same type.
https://github.com/karpathy/llm.c/blob/master/train_gpt2.c#L36

Reply:
The vocab size is 50257. It fits into int32. (I think uint32 or uint16 would be fine.)